File: reference/typescript/src/transport/sse.js
import http from "node:http";
import { Transport } from "./base.js";

export class SseServerTransport extends Transport {
  #server = null;
  #clients = new Map();
  #nextClientId = 0;

  constructor(options = {}) {
    super();
    this.port = options.port ?? 0;
    this.host = options.host ?? "127.0.0.1";
    this.path = options.path ?? "/aep/events";
    this.heartbeatInterval = options.heartbeatInterval ?? 15000;
    this.#nextClientId = 0;
  }

  async _onStart() {
    this.#server = http.createServer((req, res) => {
      const url = new URL(req.url, `http://${req.headers.host}`);

      if (req.method === "GET" && url.pathname === this.path) {
        return this.#handleStream(req, res);
      }

      if (req.method === "POST" && url.pathname === this.path) {
        return this.#handleIngest(req, res);
      }

      res.writeHead(404);
      res.end("not found");
    });

    return new Promise((resolve) => {
      this.#server.listen(this.port, this.host, () => {
        const addr = this.#server.address();
        this.port = addr.port;
        resolve();
      });
    });
  }

  async _onStop() {
    for (const [, { res }] of this.#clients) {
      try { res.end(); } catch {}
      try { res.destroy(); } catch {}
    }
    this.#clients.clear();

    if (this.#server) {
      // Destroy all active connections first, then close
      if (typeof this.#server.closeAllConnections === "function") {
        this.#server.closeAllConnections();
      }
      this.#server.unref?.();

      await new Promise((resolve) => {
        this.#server.close((err) => {
          // Resolve even on error - the server may already be closing
          resolve();
        });
        // Force stop if close doesn't fire within 1s
        setTimeout(resolve, 1000);
      });
      this.#server = null;
    }
  }

  _onSend(data) {
    const json = typeof data === "string" ? data : JSON.stringify(data);
    let event;

    try {
      event = typeof data === "string" ? JSON.parse(data) : data;
    } catch {
      event = data;
    }

    for (const [id, { res }] of this.#clients) {
      res.write(`id: ${event.id}\n`);
      if (event.type) res.write(`event: ${event.type}\n`);
      res.write(`data: ${json}\n\n`);
    }
  }

  #handleStream(req, res) {
    const clientId = ++this.#nextClientId;
    const lastEventId = req.headers["last-event-id"] ?? null;

    res.writeHead(200, {
      "Content-Type": "text/event-stream",
      "Cache-Control": "no-cache",
      Connection: "keep-alive",
      "Access-Control-Allow-Origin": "*"
    });

    this.#clients.set(clientId, { res, lastEventId });

    this.emit("connection", { clientId, lastEventId, remoteAddress: req.socket.remoteAddress });

    const heartbeat = this.heartbeatInterval > 0
      ? setInterval(() => res.write(": heartbeat\n\n"), this.heartbeatInterval)
      : null;

    req.on("close", () => {
      if (heartbeat) clearInterval(heartbeat);
      this.#clients.delete(clientId);
      this.emit("disconnection", { clientId });
    });
  }

  async #handleIngest(req, res) {
    const chunks = [];
    for await (const chunk of req) {
      chunks.push(chunk);
    }
    const body = Buffer.concat(chunks).toString("utf8");
    const lines = body.split("\n").filter((l) => l.trim().length > 0);

    const accepted = [];
    const errors = [];

    for (const line of lines) {
      try {
        const event = JSON.parse(line);
        accepted.push(event);
        this.emit("message", event);
      } catch (err) {
        errors.push({ line: line.substring(0, 100), error: err.message });
      }
    }

    res.writeHead(202, {
      "Content-Type": "application/json",
      "Access-Control-Allow-Origin": "*"
    });
    res.end(JSON.stringify({ accepted: accepted.length, rejected: errors.length, errors }));

    if (req.method === "OPTIONS") {
      res.writeHead(204, {
        "Access-Control-Allow-Origin": "*",
        "Access-Control-Allow-Methods": "POST, GET, OPTIONS",
        "Access-Control-Allow-Headers": "Content-Type, Last-Event-ID"
      });
      res.end();
    }
  }
}

export class SseClientTransport extends Transport {
  #url = null;
  #baseUrl = null;
  #abortController = null;
  #lastEventId = null;

  constructor(options = {}) {
    super();
    this.baseUrl = options.baseUrl ?? "http://127.0.0.1:0";
    this.path = options.path ?? "/aep/events";
    this.#url = `${this.baseUrl}${this.path}`;
    this.reconnectDelay = options.reconnectDelay ?? 1000;
  }

  async _onStart() {
    this.#connect();
  }

  async _onStop() {
    if (this.#abortController) {
      this.#abortController.abort();
      this.#abortController = null;
    }
  }

  async _onSend(data) {
    const body = typeof data === "string" ? data : JSON.stringify(data);

    try {
      await fetch(`${this.#url}${this.#url.includes("?") ? "&" : "?"}session_id=local`, {
        method: "POST",
        headers: { "Content-Type": "application/x-ndjson" },
        body
      });
    } catch (err) {
      if (err.name !== "AbortError") {
        this.emit("error", err);
      }
    }
  }

  async #connect() {
    this.#abortController = new AbortController();

    try {
      const headers = {
        Accept: "text/event-stream",
        "Cache-Control": "no-cache"
      };
      if (this.#lastEventId) {
        headers["Last-Event-ID"] = this.#lastEventId;
      }

      const response = await fetch(this.#url, {
        headers,
        signal: this.#abortController.signal
      });

      if (!response.ok) {
        throw new Error(`SSE connection failed: ${response.status}`);
      }

      this.emit("open");

      const reader = response.body.getReader();
      const decoder = new TextDecoder();
      let buffer = "";

      while (true) {
        const { done, value } = await reader.read();
        if (done) break;

        buffer += decoder.decode(value, { stream: true });
        const messages = buffer.split("\n\n");
        buffer = messages.pop();

        for (const msg of messages) {
          const event = this.#parseSseMessage(msg);
          if (event) {
            this.#lastEventId = event.id;
            this.emit("message", event);
          }
        }
      }
    } catch (err) {
      if (err.name !== "AbortError") {
        this.emit("error", err);
      }
    }

    this.emit("close");
  }

  #parseSseMessage(text) {
    const lines = text.split("\n");
    let id = null;
    let eventType = null;
    const dataLines = [];

    for (const line of lines) {
      if (line.startsWith("id:")) {
        id = line.slice(3).trim();
      } else if (line.startsWith("event:")) {
        eventType = line.slice(6).trim();
      } else if (line.startsWith("data:")) {
        dataLines.push(line.slice(5).trim());
      } else if (line.startsWith(":")) {
        continue;
      }
    }

    if (dataLines.length === 0) return null;

    try {
      const data = JSON.parse(dataLines.join("\n"));
      if (id) data._sseId = id;
      if (eventType) data._sseEvent = eventType;
      return data;
    } catch {
      return { type: eventType ?? "message", id, payload: { raw: dataLines.join("\n") } };
    }
  }
}

File: implementations/python/src/aep/transport/sse.py
import json
import socket
import threading
from http.server import HTTPServer, BaseHTTPRequestHandler


class _SseRequestHandler(BaseHTTPRequestHandler):
    def do_GET(self):
        if self.path == "/aep/events" or self.path.startswith("/aep/events?"):
            self._handle_stream()
        else:
            self.send_response(404)
            self.end_headers()
            self.wfile.write(b"not found")

    def do_POST(self):
        if self.path == "/aep/events" or self.path.startswith("/aep/events?"):
            self._handle_ingest()
        else:
            self.send_response(404)
            self.end_headers()
            self.wfile.write(b"not found")

    def do_OPTIONS(self):
        self.send_response(204)
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "POST, GET, OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "Content-Type, Last-Event-ID")
        self.end_headers()

    def _handle_stream(self):
        transport = self.server._aep_transport
        client_id = transport._next_client_id
        transport._next_client_id += 1

        self.send_response(200)
        self.send_header("Content-Type", "text/event-stream")
        self.send_header("Cache-Control", "no-cache")
        self.send_header("Connection", "keep-alive")
        self.send_header("Access-Control-Allow-Origin", "*")
        self.end_headers()

        client_entry = {"wfile": self.wfile, "client_id": client_id}
        transport._clients[client_id] = client_entry

        transport.emit("connection", {"clientId": client_id})

        heartbeat = None
        if transport._heartbeat_interval > 0:
            heartbeat = threading.Timer(
                transport._heartbeat_interval / 1000.0,
                transport._send_heartbeat,
                args=[client_id]
            )
            heartbeat.daemon = True
            heartbeat.start()

        try:
            while True:
                line = self.rfile.readline()
                if not line:
                    break
        except (ConnectionError, OSError):
            pass
        finally:
            if heartbeat:
                heartbeat.cancel()
            transport._clients.pop(client_id, None)
            transport.emit("disconnection", {"clientId": client_id})

    def _handle_ingest(self):
        transport = self.server._aep_transport
        content_length = int(self.headers.get("Content-Length", 0))
        if content_length > 0:
            body = self.rfile.read(content_length).decode("utf-8")
        else:
            body = ""

        lines = [line for line in body.split("\n") if line.strip()]
        accepted = []
        errors = []

        for line in lines:
            try:
                event = json.loads(line)
                accepted.append(event)
                transport.emit("message", event)
            except json.JSONDecodeError as err:
                errors.append({"line": line[:100], "error": str(err)})

        self.send_response(202)
        self.send_header("Content-Type", "application/json")
        self.send_header("Access-Control-Allow-Origin", "*")
        self.end_headers()
        response = json.dumps({
            "accepted": len(accepted),
            "rejected": len(errors),
            "errors": errors,
        })
        self.wfile.write(response.encode("utf-8"))

    def log_message(self, format, *args):
        pass


class SseServerTransport:
    def __init__(self, options=None, **kwargs):
        if options is None:
            options = kwargs
        if not options:
            options = {}
        self.port = options.get("port", 0)
        self.host = options.get("host", "127.0.0.1")
        self.path = options.get("path", "/aep/events")
        self._heartbeat_interval = options.get("heartbeatInterval", 15000)

        self._listeners = {}
        self._clients = {}
        self._next_client_id = 0
        self._server = None
        self._lock = threading.Lock()

    def on(self, event, callback):
        with self._lock:
            if event not in self._listeners:
                self._listeners[event] = []
            self._listeners[event].append(callback)

    def emit(self, event, data=None):
        with self._lock:
            callbacks = list(self._listeners.get(event, []))
        for cb in callbacks:
            try:
                cb(data)
            except Exception:
                pass

    def start(self):
        self._server = HTTPServer((self.host, self.port), _SseRequestHandler)
        self._server._aep_transport = self
        self.port = self._server.server_port
        self._server.serve_forever(poll_interval=0.1)

    def stop(self):
        if self._server:
            self._server.shutdown()
            self._server.server_close()
            self._server = None

    def send(self, data):
        if isinstance(data, str):
            data = json.loads(data)
        json_str = json.dumps(data)
        event_id = data.get("id", "")
        event_type = data.get("type", "")

        dead_clients = []
        with self._lock:
            clients = list(self._clients.items())

        for client_id, client in clients:
            try:
                wfile = client["wfile"]
                if event_id:
                    wfile.write(f"id: {event_id}\n".encode("utf-8"))
                if event_type:
                    wfile.write(f"event: {event_type}\n".encode("utf-8"))
                wfile.write(f"data: {json_str}\n\n".encode("utf-8"))
                wfile.flush()
            except Exception:
                dead_clients.append(client_id)

        for client_id in dead_clients:
            self._clients.pop(client_id, None)

    def _send_heartbeat(self, client_id):
        client = self._clients.get(client_id)
        if client is None:
            return
        try:
            client["wfile"].write(b": heartbeat\n\n")
            client["wfile"].flush()
        except Exception:
            self._clients.pop(client_id, None)
            return

        if self._heartbeat_interval > 0:
            heartbeat = threading.Timer(
                self._heartbeat_interval / 1000.0,
                self._send_heartbeat,
                args=[client_id]
            )
            heartbeat.daemon = True
            heartbeat.start()
